File: java/construct_binary_tree_from_inorder_and_postorder_traversal_106/Solution.java
package construct_binary_tree_from_inorder_and_postorder_traversal_106;

import tree.binary_tree.TreeNode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
  public TreeNode buildTree(int[] inorder, int[] postorder) {
    List<Integer> postorderList = new ArrayList<>();
    for (int num : postorder) {
      postorderList.add(num);
    }

    return build(inorder, postorderList);
  }

  private static TreeNode build(int[] inorder, List<Integer> postorderList) {
    if (inorder.length == 0) {
      return null;
    }

    int num = postorderList.removeLast();
    int index = indexOf(inorder, num);

    TreeNode root = new TreeNode(num);
    root.right = build(Arrays.copyOfRange(inorder, index + 1, inorder.length), postorderList);
    root.left = build(Arrays.copyOfRange(inorder, 0, index), postorderList);

    return root;
  }

  private static int indexOf(int[] array, int target) {
    for (int i = 0; i < array.length; i++) {
      if (array[i] == target) {
        return i;
      }
    }
    return -1;
  }
}

File: python/construct_binary_tree_from_inorder_and_postorder_traversal_106/Solution.py
from typing import List, Optional, Self

from tree.binary_tree.TreeNode import TreeNode


class Solution:
    def buildTree(
        self: Self, inorder: List[int], postorder: List[int]
    ) -> Optional[TreeNode]:
        if len(inorder) == 0:
            return None

        num: int = postorder.pop()
        index: int = inorder.index(num)

        root = TreeNode(num)
        root.right = self.buildTree(inorder[index + 1 :], postorder)
        root.left = self.buildTree(inorder[:index], postorder)

        return root
